test/distributed/org/apache/cassandra/distributed/test/ShortReadProtectionTest.java

```diff
@@ -18,46 +18,309 @@
 
 package org.apache.cassandra.distributed.test;
 
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+import java.util.stream.IntStream;
 
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.BeforeClass;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 
+import org.apache.cassandra.dht.Murmur3Partitioner;
 import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.ConsistencyLevel;
 import org.apache.cassandra.distributed.api.ICoordinator;
-import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.shared.AssertUtils;
+import org.apache.cassandra.utils.ByteBufferUtil;
 
-import static java.lang.String.format;
-import static java.util.Arrays.asList;
+import static com.google.common.collect.Iterators.toArray;
 import static org.apache.cassandra.distributed.api.ConsistencyLevel.ALL;
-import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.QUORUM;
+import static org.apache.cassandra.distributed.shared.AssertUtils.row;
 
 /**
  * Tests short read protection, the mechanism that ensures distributed queries at read consistency levels > ONE/LOCAL_ONE
  * avoid short reads that might happen when a limit is used and reconciliation accepts less rows than such limit.
  */
+@RunWith(Parameterized.class)
 public class ShortReadProtectionTest extends TestBaseImpl
 {
+    private static final int NUM_NODES = 3;
+
+    private static Cluster cluster;
+    private Tester tester;
+
+    /**
+     * {@code true} for CL=QUORUM writes and CL=QUORUM reads, {@code false} for CL=ONE writes and CL=ALL reads.
+     */
+    @Parameterized.Parameter
+    public boolean quorum;
+
+    /**
+     * Whether to flush data after mutations.
+     */
+    @Parameterized.Parameter(1)
+    public boolean flush;
+
+    /**
+     * Whether paging is used for the distributed queries.
+     */
+    @Parameterized.Parameter(2)
+    public boolean paging;
```

Comment:
@adelapena Do any of the tests actually use a fetch size that would cause more than a single page to be processed?

Reply:
I think yes, at least all the tests returning more than one row should require more that one page:
* testSkinnyTableWithLiveRows
* testMultipleMissedRows
* testAscendingOrder
* testDescendingOrder
* testSkipEarlyTerminationRows
* testSkipEarlyTerminationPartitions

Also, even if there is only zero or one results to return and a page size of one, paging is still interesting with a limit greater than one because it can expose failures like CASSANDRA-16307. 

I have added [here](https://github.com/apache/cassandra/pull/801/commits/c77a8870e8865a322706a8b49d69f264f5ed2dcd) some additional page sizes, just in case. These page sizes are not added as a JUnit parameter to not make the tests much slower.

Reply:
I initially missed that the page size was hard-coded to 1 in `assertRows()`. In any case, this looks good. I'd honestly be fine w/ fetch sizes of just 1 and 10, given the kinds of results we get here.

Reply:
Right, I have reduced the number of fetch sizes. Also, the query without paging was needlessly done inside the  `PAGE_SIZES` loop; I have moved it out. 
Also, I have moved the selection of coordinator from the JUnit parameters to the inside of the `assertRows` method. This should be valid since read repair strategy is disabled, and it makes the tests significantly faster (from 1m42s to 52s in my laptop).